lib/charms/kratos_external_idp_integrator/v1/kratos_external_provider.py

```diff
@@ -0,0 +1,604 @@
+        logger.info(f"The client's redirect_uri changed to {event.redirect_uri}")
+
+    def _on_ready(self, event):
+        if not isinstance(self.unit.status, BlockedStatus):
+            self.external_idp_provider.create_provider(providers)
+```
+
+To use the library from the requirer side (Kratos):
+
+In the `metadata.yaml` of the charm, add the following:
+```yaml
+requires:
+    kratos-external-idp:
+        interface: external_provider
+```
+
+Then, to initialize the library:
+
+```python
+from charms.kratos_external_idp_integrator.v1.kratos_external_provider import (
+    ExternalIdpRequirer
+)
+
+class KratosCharm(CharmBase):
+  def __init__(self, *args):
+    # ...
+    self.external_idp_requirer = ExternalIdpRequirer(self)
+
+    self.framework.observe(
+        self.external_idp_provider.on.client_config_changed, self._on_client_config_changed
+    )
+
+    def _on_client_config_changed(self, event):
+        self._configure(event)
+
+        self.external_provider.update_registered_provider(
+            providers,
+            event.relation_id,
+        )
+```
+"""
+
+import base64
+import hashlib
+import json
+import logging
+from typing import Annotated, Iterator, Literal, Mapping, Union
+
+from ops.charm import (
+    CharmBase,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationDepartedEvent,
+    RelationEvent,
+    RelationJoinedEvent,
+)
+from ops.framework import EventBase, EventSource, Handle, Object, ObjectEvents
+from ops.model import Relation, TooManyRelatedAppsError
+from pydantic import (
+    AliasChoices,
+    BaseModel,
+    Field,
+    RootModel,
+    SecretStr,
+    SerializerFunctionWrapHandler,
+    ValidationError,
+    WrapSerializer,
+    field_serializer,
+    field_validator,
+    model_validator,
+)
+from typing_extensions import Self
+
+# The unique CharmHub library identifier, never change it
+LIBID = "33040051de7f43a8bb43349f2b037dfc"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 1
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 0
+
+PYDEPS = ["pydantic~=2.11"]
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "kratos-external-idp"
+ALLOWED_PROVIDERS = {
+    "generic",
+    "google",
+    "facebook",
+    "microsoft",
+    "github",
+    "apple",
+    "gitlab",
+    "auth0",
+    "slack",
+    "spotify",
+    "discord",
+    "twitch",
+    "netid",
+    "yander",
+    "vk",
+    "dingtalk",
+}
+
+
+def dump_secret(v: SecretStr, _: SerializerFunctionWrapHandler) -> str:
+    return v.get_secret_value()
+
+
+Secret = Annotated[SecretStr, WrapSerializer(dump_secret)]
+
+
+class BaseProvider(BaseModel):
+    provider: str
+    client_id: str
+    scope: list[str] = ["profile", "email", "address", "phone"]
+    secret_backend: Literal["relation", "secret", "vault"] = Field(default="relation")
+    id: str | None = Field(
```

Comment:
nit (cosmetic): Any reason not to use `Optional` instead?

```suggestion
    id: Optional[str] = Field(
```